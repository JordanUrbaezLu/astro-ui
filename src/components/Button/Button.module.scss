@use "../../utility/globals";

.button {
  align-items: center;
  border: 2px solid globals.$primary-background-color;
  cursor: pointer;
  display: flex;
  outline: none;
}

.primary {
  background-color: globals.$primary-background-color;

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
    border: 2px solid
      globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
    border: 2px solid globals.$primary-background-color-active;
  }
}

.secondary {
  background-color: white;
  border: 2px solid globals.$secondary-text-color;

  &:hover,
  &:focus {
    background-color: globals.$secondary-background-color-hover-and-active;
    border: 2px solid globals.$secondary-border-color-hover-and-focus;
    color: globals.$secondary-text-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$secondary-background-color-active;
  }
}

.small {
  border-radius: 6px;
  height: 25.8px;
  padding: 4px 6px;

  .leadingIcon,
  .trailingIcon {
    height: 12px;
  }
}

.medium {
  border-radius: 8px;
  height: 34.5px;
  padding: 6px 9px;

  .leadingIcon,
  .trailingIcon {
    height: 16px;
  }
}

.large {
  border-radius: 10px;
  height: 43px;
  padding: 8px 12px;

  .leadingIcon,
  .trailingIcon {
    height: 20px;
  }
}

.smallRipple {
  border-radius: 6px;
}

.mediumRipple {
  border-radius: 8px;
}

.largeRipple {
  border-radius: 10px;
}

.typography {
  align-items: center;
  display: flex;
  gap: 6px;
}

.disableRipple {
  pointer-events: none !important;
}

.disabled {
  background-color: lightgray !important;
  border: 2px solid lightgray;
  cursor: default !important;
  user-select: none !important;
}
