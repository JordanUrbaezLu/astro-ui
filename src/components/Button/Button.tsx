import * as React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";
import { Typography } from "../Typography/Typography";
import { Ripple } from "../Ripple";

export type ButtonVariant = "primary" | "secondary";
export type ButtonSize = "small" | "medium" | "large";
export type ButtonType = "button" | "submit" | "reset";

export interface ButtonProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The content for the Button
   */
  children: string;
  /**
   * If the Button is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * The leading icon for the Button
   */
  leadingIcon?: React.ReactNode;
  /**
   * The callback fired when the Button is clicked
   */
  onClick?: () => void;
  /**
   * The size of the Button
   *
   * @default "medium"
   */
  size?: ButtonSize;
  /**
   * The trailing icon for the Button
   */
  trailingIcon?: React.ReactNode;
  /**
   * The type of the Button
   */
  type?: ButtonType;
  /**
   * The variant of the Button
   *
   * @default "primary"
   */
  variant?: ButtonVariant;
}

/**
 * @public
 *
 * @description
 *
 * The Button component is used to allows user to take an action.
 */
export const Button = React.forwardRef<
  HTMLButtonElement,
  ButtonProps
>((props, ref) => {
  const {
    className,
    children,
    disabled = false,
    leadingIcon,
    onClick,
    size = "medium",
    trailingIcon,
    type,
    variant = "primary",
    ...rest
  } = props;

  return (
    <button
      className={classNames(
        className,
        styles.button,
        disabled && styles.disabled,
        size === "small"
          ? styles.small
          : size === "large"
          ? styles.large
          : styles.medium,
        variant === "secondary" ? styles.secondary : styles.primary
      )}
      disabled={disabled}
      onClick={onClick}
      ref={ref}
      type={
        type === "reset"
          ? "reset"
          : type === "submit"
          ? "submit"
          : "button"
      }
      {...rest}
    >
      {!disabled && <Ripple position="cursor" />}
      <Typography
        className={styles.typography}
        type={
          size === "large" ? "h1" : size === "medium" ? "h2" : "h3"
        }
        variant={variant}
      >
        {leadingIcon && (
          <span className={styles.leadingIcon}>{leadingIcon}</span>
        )}
        {children}
        {trailingIcon && (
          <span className={styles.trailingIcon}>{trailingIcon}</span>
        )}
      </Typography>
    </button>
  );
});

Button.displayName = "Button";
