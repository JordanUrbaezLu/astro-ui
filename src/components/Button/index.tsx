/* istanbul ignore file */
export * from "./Button";
