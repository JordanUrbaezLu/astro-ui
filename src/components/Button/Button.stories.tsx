import { Story } from "@storybook/react";
import { Button, ButtonProps } from "./Button";
import {
  MdOutlineArrowBack,
  MdOutlineArrowForward,
} from "react-icons/md";
import { FlexRowContainer } from "../../../.storybook/components/FlexRowContainer";

export default {
  component: Button,
  title: "Components/Inputs/Button",
};

const Template: Story<ButtonProps> = (args) => {
  return <Button {...args}>Button</Button>;
};

export const Default = Template.bind({});

export const Size = () => {
  return (
    <FlexRowContainer>
      <Button size="small">Small</Button>
      <Button size="medium">Medium</Button>
      <Button size="large">Large</Button>
    </FlexRowContainer>
  );
};

export const Variant = () => {
  return (
    <FlexRowContainer>
      <Button variant="primary">Primary</Button>
      <Button variant="secondary">Secondary</Button>
    </FlexRowContainer>
  );
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const LeadingIcon = Template.bind({});
LeadingIcon.args = {
  leadingIcon: <MdOutlineArrowBack />,
};

export const TrailingIcon = Template.bind({});
TrailingIcon.args = {
  trailingIcon: <MdOutlineArrowForward />,
};

export const LeadingIconAndTrailingIcon = Template.bind({});
LeadingIconAndTrailingIcon.args = {
  leadingIcon: <MdOutlineArrowBack />,
  trailingIcon: <MdOutlineArrowForward />,
};

export const AllProps = Template.bind({});
AllProps.args = {
  size: "large",
  variant: "secondary",
  leadingIcon: <MdOutlineArrowBack />,
  trailingIcon: <MdOutlineArrowForward />,
};
