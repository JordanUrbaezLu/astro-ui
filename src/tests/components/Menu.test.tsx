import {
  createEvent,
  fireEvent,
  render,
  screen,
} from "@testing-library/react";
import {
  Menu,
  MenuAlignment,
  MenuSeparator,
  MenuItem,
  MenuHeader,
} from "../../components/Menu";
import { axe } from "jest-axe";
import { Button } from "../../components/Button/Button";
import { act } from "react-dom/test-utils";

// Disable console.error for each test in this file
// console.error Warning: `NaN` is an invalid value for the `left` css style property. (Ripples Click)
beforeEach(() => {
  jest.spyOn(console, "error").mockImplementation(() => undefined);
});

describe("Menu", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Menu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Menu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should trigger onOpen when trigger is clicked.", () => {
    const onOpen = jest.fn();

    render(
      <Menu
        isOpen={false}
        onOpen={onOpen}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );

    act(() => screen.getByRole("button").click());

    expect(onOpen).toHaveBeenCalledTimes(1);
  });

  test("Should trigger onClose when trigger is clicked.", () => {
    const onClose = jest.fn();

    render(
      <Menu
        isOpen
        onOpen={jest.fn()}
        onClose={onClose}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );

    act(() => screen.getByRole("button").click());

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should trigger onClose on Space on a Menu Item.", () => {
    const onClose = jest.fn();

    render(
      <Menu
        isOpen
        onOpen={jest.fn()}
        onClose={onClose}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );

    const firstMenuItem = screen.getAllByRole("menuitem")[0];

    const event = createEvent.keyDown(firstMenuItem, {
      code: "Space",
    });

    fireEvent(firstMenuItem, event);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should trigger onClose on Escape key down.", () => {
    const onClose = jest.fn();

    render(
      <Menu
        isOpen
        onOpen={jest.fn()}
        onClose={onClose}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );

    fireEvent.keyDown(document.body, { key: "Escape" });

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should trigger onClose when Menu Item is clicked.", () => {
    const onClose = jest.fn();

    render(
      <Menu
        isOpen
        onOpen={jest.fn()}
        onClose={onClose}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );

    const firstMenuItem = screen.getAllByRole("menuitem")[0];

    firstMenuItem.click();

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("Should handle focus correctly", () => {
    render(
      <Menu
        isOpen
        onOpen={jest.fn()}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
        <MenuSeparator />
        <MenuItem>Two</MenuItem>
        <MenuHeader>Header</MenuHeader>
        <MenuItem>Three</MenuItem>
      </Menu>
    );

    const firstMenuItem = screen.getAllByRole("menuitem")[0];

    expect(firstMenuItem).toHaveFocus();

    const event1 = createEvent.keyDown(firstMenuItem, {
      code: "ArrowDown",
    });

    fireEvent(firstMenuItem, event1);

    const secondMenuItem = screen.getAllByRole("menuitem")[1];

    expect(secondMenuItem).toHaveFocus();

    const event2 = createEvent.keyDown(secondMenuItem, {
      code: "ArrowDown",
    });

    fireEvent(secondMenuItem, event2);

    const thirdMenuItem = screen.getAllByRole("menuitem")[2];

    expect(thirdMenuItem).toHaveFocus();

    const event3 = createEvent.keyDown(thirdMenuItem, {
      code: "ArrowDown",
    });

    fireEvent(thirdMenuItem, event3);

    expect(firstMenuItem).toHaveFocus();

    const event4 = createEvent.keyDown(firstMenuItem, {
      code: "ArrowUp",
    });

    fireEvent(firstMenuItem, event4);

    expect(thirdMenuItem).toHaveFocus();

    const event5 = createEvent.keyDown(thirdMenuItem, {
      code: "ArrowUp",
    });

    fireEvent(thirdMenuItem, event5);

    expect(secondMenuItem).toHaveFocus();
  });
});

describe("Props", () => {
  test("Should render children correctly.", () => {
    render(
      <Menu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    fireEvent.click(screen.getByRole("button"));
    expect(screen.getByRole("menuitem")).toHaveTextContent("One");
  });

  test("Should render trigger correctly.", () => {
    render(
      <Menu
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    expect(screen.getAllByRole("button")[0]).toHaveTextContent(
      "Trigger"
    );
  });

  test("Should render alignment correctly.", () => {
    render(
      <Menu
        alignment="topLeft"
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    expect(screen.getByRole("menu")).toBeInTheDocument();
  });

  test("Should render closed correctly.", () => {
    render(
      <Menu
        alignment="topLeft"
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    expect(screen.queryByRole("menu")).not.toBeInTheDocument();
  });

  test.each<MenuAlignment>([
    "bottomLeft",
    "bottomCenter",
    "bottomRight",
    "topLeft",
    "topCenter",
    "topRight",
  ])("Should render alignment correctly.", (alignment) => {
    render(
      <Menu
        alignment={alignment}
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
      >
        <MenuItem>One</MenuItem>
      </Menu>
    );
    expect(screen.getByRole("menuitem")).toBeInTheDocument();
  });
});
