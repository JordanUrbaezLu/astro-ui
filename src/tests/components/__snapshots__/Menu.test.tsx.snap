// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Should render component correctly. 1`] = `
<span
  class="bottom container"
>
  <button
    class="isOpen trigger button medium primary"
    role="button"
    tabindex="0"
    type="button"
  >
    <div
      class="container"
      role="presentation"
    />
    <div
      class="typography typography typographyh2 primary normal"
    >
      Trigger
    </div>
  </button>
  <div
    class="bottom layoutContainer"
    style="transform: translate3d(0px, 2px, 0);"
  >
    <div
      class="baseMenu"
    >
      <div
        role="menu"
      >
        <div
          class="menuItemContainer typography typographyh2 primary normal"
          role="menuitem"
          tabindex="0"
        >
          One
        </div>
      </div>
    </div>
  </div>
</span>
`;
